File: client-go/queue_builder.go
package queen

import (
	"context"
	"fmt"
	"net/url"
	"strconv"
)

// QueueBuilder provides a fluent API for queue operations.
type QueueBuilder struct {
	queen     *Queen
	queueName string

	// Configuration
	namespace        string
	task             string
	partition        string
	queueConfig      *QueueConfig
	bufferConfig     *BufferConfig
	consumerGroup    string
	concurrency      int
	batch            int
	limit            int
	idleMillis       int
	autoAck          *bool
	wait             *bool
	timeoutMillis    int
	renewLease       bool
	renewLeaseMillis int
	subscriptionMode string
	subscriptionFrom string
	each             bool
}

// NewQueueBuilder creates a new QueueBuilder.
func NewQueueBuilder(queen *Queen, queueName string) *QueueBuilder {
	return &QueueBuilder{
		queen:     queen,
		queueName: queueName,
		partition: DefaultPartition,
	}
}

// Namespace sets the namespace for the queue.
func (qb *QueueBuilder) Namespace(name string) *QueueBuilder {
	qb.namespace = name
	return qb
}

// Task sets the task name for the queue.
func (qb *QueueBuilder) Task(name string) *QueueBuilder {
	qb.task = name
	return qb
}

// Partition sets the partition for operations.
func (qb *QueueBuilder) Partition(name string) *QueueBuilder {
	if name == "" {
		qb.partition = DefaultPartition
	} else {
		qb.partition = name
	}
	return qb
}

// Config sets the queue configuration for create operations.
func (qb *QueueBuilder) Config(config QueueConfig) *QueueBuilder {
	qb.queueConfig = &config
	return qb
}

// Buffer enables client-side buffering for push operations.
func (qb *QueueBuilder) Buffer(config BufferConfig) *QueueBuilder {
	qb.bufferConfig = &config
	return qb
}

// Group sets the consumer group for consume/pop operations.
func (qb *QueueBuilder) Group(name string) *QueueBuilder {
	qb.consumerGroup = name
	return qb
}

// Concurrency sets the number of concurrent workers for consume operations.
func (qb *QueueBuilder) Concurrency(count int) *QueueBuilder {
	qb.concurrency = count
	return qb
}

// Batch sets the batch size for consume/pop operations.
func (qb *QueueBuilder) Batch(size int) *QueueBuilder {
	qb.batch = size
	return qb
}

// Limit sets the maximum number of messages to process.
func (qb *QueueBuilder) Limit(count int) *QueueBuilder {
	qb.limit = count
	return qb
}

// IdleMillis sets the idle timeout in milliseconds.
func (qb *QueueBuilder) IdleMillis(millis int) *QueueBuilder {
	qb.idleMillis = millis
	return qb
}

// AutoAck sets whether to automatically acknowledge messages.
func (qb *QueueBuilder) AutoAck(enabled bool) *QueueBuilder {
	qb.autoAck = &enabled
	return qb
}

// Wait sets whether to use long polling.
func (qb *QueueBuilder) Wait(enabled bool) *QueueBuilder {
	qb.wait = &enabled
	return qb
}

// TimeoutMillis sets the timeout for long polling.
func (qb *QueueBuilder) TimeoutMillis(millis int) *QueueBuilder {
	qb.timeoutMillis = millis
	return qb
}

// RenewLease enables automatic lease renewal during processing.
func (qb *QueueBuilder) RenewLease(enabled bool, intervalMillis int) *QueueBuilder {
	qb.renewLease = enabled
	qb.renewLeaseMillis = intervalMillis
	return qb
}

// SubscriptionMode sets the subscription mode (all, new, new-only).
func (qb *QueueBuilder) SubscriptionMode(mode string) *QueueBuilder {
	qb.subscriptionMode = mode
	return qb
}

// SubscriptionFrom sets the subscription start point (now, timestamp).
func (qb *QueueBuilder) SubscriptionFrom(from string) *QueueBuilder {
	qb.subscriptionFrom = from
	return qb
}

// Each sets whether to process messages one at a time.
func (qb *QueueBuilder) Each() *QueueBuilder {
	qb.each = true
	return qb
}

// Create returns an OperationBuilder for creating the queue.
func (qb *QueueBuilder) Create() *OperationBuilder {
	return NewOperationBuilder(qb, "create")
}

// Delete returns an OperationBuilder for deleting the queue.
func (qb *QueueBuilder) Delete() *OperationBuilder {
	return NewOperationBuilder(qb, "delete")
}

// Push returns a PushBuilder for pushing messages.
func (qb *QueueBuilder) Push(payload interface{}) *PushBuilder {
	return NewPushBuilder(qb, payload)
}

// Pop pops messages from the queue.
func (qb *QueueBuilder) Pop(ctx context.Context) ([]*Message, error) {
	// Validate queue name
	if qb.queueName == "" && qb.namespace == "" && qb.task == "" {
		return nil, fmt.Errorf("queue name, namespace, or task is required")
	}

	// Build path
	var path string
	if qb.queueName != "" {
		if qb.partition != "" && qb.partition != DefaultPartition {
			path = fmt.Sprintf("/api/v1/pop/queue/%s/partition/%s", url.PathEscape(qb.queueName), url.PathEscape(qb.partition))
		} else {
			path = fmt.Sprintf("/api/v1/pop/queue/%s", url.PathEscape(qb.queueName))
		}
	} else {
		path = "/api/v1/pop"
	}

	// Build query params
	params := qb.buildPopParams()
	if params != "" {
		path += "?" + params
	}

	// Determine timeout
	timeout := qb.timeoutMillis
	if timeout == 0 {
		timeout = PopDefaults.TimeoutMillis
	}

	// Add buffer for long polling
	wait := qb.wait
	if wait == nil {
		w := PopDefaults.Wait
		wait = &w
	}
	if *wait {
		timeout += 5000 // Add 5s buffer for long polling
	}

	// Make request
	result, err := qb.queen.httpClient.Get(ctx, path, timeout, "")
	if err != nil {
		return nil, fmt.Errorf("pop request failed: %w", err)
	}

	// Parse response
	messages := parseMessages(result)

	logDebug("QueueBuilder.Pop", map[string]interface{}{
		"queue":   qb.queueName,
		"count":   len(messages),
	})

	return messages, nil
}

// buildPopParams builds the query parameters for pop requests.
func (qb *QueueBuilder) buildPopParams() string {
	params := url.Values{}

	// Batch size
	batch := qb.batch
	if batch == 0 {
		batch = PopDefaults.Batch
	}
	params.Set("batch", strconv.Itoa(batch))

	// Wait (long polling)
	wait := qb.wait
	if wait == nil {
		w := PopDefaults.Wait
		wait = &w
	}
	params.Set("wait", strconv.FormatBool(*wait))

	// Timeout
	timeout := qb.timeoutMillis
	if timeout == 0 {
		timeout = PopDefaults.TimeoutMillis
	}
	params.Set("timeout", strconv.Itoa(timeout))

	// Consumer group
	if qb.consumerGroup != "" {
		params.Set("consumerGroup", qb.consumerGroup)
	}

	// Auto ack (for pop, this is server-side)
	if qb.autoAck != nil {
		params.Set("autoAck", strconv.FormatBool(*qb.autoAck))
	}

	// Subscription mode
	if qb.subscriptionMode != "" {
		params.Set("subscriptionMode", qb.subscriptionMode)
	}

	// Subscription from
	if qb.subscriptionFrom != "" {
		params.Set("subscriptionFrom", qb.subscriptionFrom)
	}

	// Namespace and task (for namespace/task mode)
	if qb.namespace != "" {
		params.Set("namespace", qb.namespace)
	}
	if qb.task != "" {
		params.Set("task", qb.task)
	}

	return params.Encode()
}

// Consume starts consuming messages from the queue.
func (qb *QueueBuilder) Consume(ctx context.Context, handler MessageHandler) *ConsumeBuilder {
	return NewConsumeBuilder(qb, handler)
}

// ConsumeBatch starts consuming messages in batches.
func (qb *QueueBuilder) ConsumeBatch(ctx context.Context, handler BatchMessageHandler) *ConsumeBuilder {
	return NewConsumeBatchBuilder(qb, handler)
}

// FlushBuffer flushes the buffer for this queue/partition.
func (qb *QueueBuilder) FlushBuffer(ctx context.Context) error {
	key := qb.getBufferKey()
	return qb.queen.bufferManager.Flush(ctx, key)
}

// DLQ returns a DLQBuilder for querying the dead letter queue.
func (qb *QueueBuilder) DLQ(consumerGroup string) *DLQBuilder {
	return NewDLQBuilder(qb.queen.httpClient, qb.queueName, consumerGroup, qb.partition)
}

// getBufferKey returns the buffer key for this queue/partition.
func (qb *QueueBuilder) getBufferKey() string {
	partition := qb.partition
	if partition == "" {
		partition = DefaultPartition
	}
	return fmt.Sprintf("%s/%s", qb.queueName, partition)
}

// getConsumeOptions returns the consume options from the builder configuration.
func (qb *QueueBuilder) getConsumeOptions() ConsumeOptions {
	opts := ConsumeOptions{
		Queue:            qb.queueName,
		Partition:        qb.partition,
		Namespace:        qb.namespace,
		Task:             qb.task,
		Group:            qb.consumerGroup,
		Concurrency:      qb.concurrency,
		Batch:            qb.batch,
		Limit:            qb.limit,
		IdleMillis:       qb.idleMillis,
		TimeoutMillis:    qb.timeoutMillis,
		RenewLease:       qb.renewLease,
		RenewLeaseIntervalMillis: qb.renewLeaseMillis,
		SubscriptionMode: qb.subscriptionMode,
		SubscriptionFrom: qb.subscriptionFrom,
		Each:             qb.each,
	}

	// Apply defaults
	if opts.Concurrency == 0 {
		opts.Concurrency = ConsumeDefaults.Concurrency
	}
	if opts.Batch == 0 {
		opts.Batch = ConsumeDefaults.Batch
	}
	if opts.TimeoutMillis == 0 {
		opts.TimeoutMillis = ConsumeDefaults.TimeoutMillis
	}

	// Auto ack
	if qb.autoAck != nil {
		opts.AutoAck = *qb.autoAck
	} else {
		opts.AutoAck = ConsumeDefaults.AutoAck
	}

	// Wait
	if qb.wait != nil {
		opts.Wait = *qb.wait
	} else {
		opts.Wait = ConsumeDefaults.Wait
	}

	return opts
}

// parseMessages parses messages from the response.
func parseMessages(result map[string]interface{}) []*Message {
	var messages []*Message

	if result == nil {
		return messages
	}

	// Get messages array
	msgsRaw, ok := result["messages"]
	if !ok {
		return messages
	}

	msgsArray, ok := msgsRaw.([]interface{})
	if !ok {
		return messages
	}

	for _, msgRaw := range msgsArray {
		msgMap, ok := msgRaw.(map[string]interface{})
		if !ok {
			continue
		}

		msg := &Message{}

		if v, ok := msgMap["transactionId"].(string); ok {
			msg.TransactionID = v
		}
		if v, ok := msgMap["partitionId"].(string); ok {
			msg.PartitionID = v
		}
		if v, ok := msgMap["leaseId"].(string); ok {
			msg.LeaseID = v
		}
		if v, ok := msgMap["queue"].(string); ok {
			msg.Queue = v
		}
		if v, ok := msgMap["partition"].(string); ok {
			msg.Partition = v
		}
		if v, ok := msgMap["data"].(map[string]interface{}); ok {
			msg.Data = v
		}
		if v, ok := msgMap["createdAt"].(string); ok {
			msg.CreatedAt = v
		}
		if v, ok := msgMap["errorMessage"].(string); ok {
			msg.ErrorMessage = v
		}
		if v, ok := msgMap["retryCount"].(float64); ok {
			msg.RetryCount = int(v)
		}

		messages = append(messages, msg)
	}

	return messages
}

File: client-laravel/src/Builders/QueueBuilder.php
<?php

namespace Queen\Builders;

use Queen\Queen;
use Queen\Http\HttpClient;
use Queen\Buffer\BufferManager;
use Queen\Consumer\HighLevelConsumer;
use Queen\Support\Defaults;
use Queen\Support\Uuid;

class QueueBuilder
{
    private Queen $queen;
    private HttpClient $httpClient;
    private BufferManager $bufferManager;
    private ?string $queueName;
    private string $partition = 'Default';
    private ?string $namespace = null;
    private ?string $task = null;
    private ?string $group = null;
    private array $config = [];

    // Consume options
    private int $consumeConcurrency;
    private int $consumeBatch;
    private ?int $consumeLimit;
    private ?int $consumeIdleMillis;
    private bool $consumeAutoAck;
    private bool $consumeWait;
    private int $consumeTimeoutMillis;
    private bool $consumeRenewLease;
    private ?int $consumeRenewLeaseIntervalMillis;
    private ?string $consumeSubscriptionMode;
    private ?string $consumeSubscriptionFrom;
    private bool $consumeEach = false;

    // Buffer options
    private ?array $bufferOptions = null;

    public function __construct(Queen $queen, HttpClient $httpClient, BufferManager $bufferManager, ?string $queueName = null)
    {
        $this->queen = $queen;
        $this->httpClient = $httpClient;
        $this->bufferManager = $bufferManager;
        $this->queueName = $queueName;

        // Initialize from consume defaults
        $d = Defaults::CONSUME_DEFAULTS;
        $this->consumeConcurrency = $d['concurrency'];
        $this->consumeBatch = $d['batch'];
        $this->consumeLimit = $d['limit'];
        $this->consumeIdleMillis = $d['idleMillis'];
        $this->consumeAutoAck = $d['autoAck'];
        $this->consumeWait = $d['wait'];
        $this->consumeTimeoutMillis = $d['timeoutMillis'];
        $this->consumeRenewLease = $d['renewLease'];
        $this->consumeRenewLeaseIntervalMillis = $d['renewLeaseIntervalMillis'];
        $this->consumeSubscriptionMode = $d['subscriptionMode'];
        $this->consumeSubscriptionFrom = $d['subscriptionFrom'];
    }

    // ===========================
    // Affinity Key Generation
    // ===========================

    private function getAffinityKey(): ?string
    {
        if ($this->queueName !== null) {
            $partition = $this->partition ?: '*';
            $group = $this->group ?: '__QUEUE_MODE__';
            return "{$this->queueName}:{$partition}:{$group}";
        }

        if ($this->namespace !== null || $this->task !== null) {
            $ns = $this->namespace ?: '*';
            $task = $this->task ?: '*';
            $group = $this->group ?: '__QUEUE_MODE__';
            return "{$ns}:{$task}:{$group}";
        }

        return null;
    }

    // ===========================
    // Queue Configuration
    // ===========================

    public function namespace(string $name): static
    {
        $this->namespace = $name;
        return $this;
    }

    public function task(string $name): static
    {
        $this->task = $name;
        return $this;
    }

    public function config(array $options): static
    {
        $this->config = array_merge(Defaults::QUEUE_DEFAULTS, $options);
        return $this;
    }

    public function create(): OperationBuilder
    {
        $fullConfig = !empty($this->config) ? $this->config : Defaults::QUEUE_DEFAULTS;

        return new OperationBuilder($this->httpClient, 'POST', '/api/v1/configure', [
            'queue' => $this->queueName,
            'namespace' => $this->namespace,
            'task' => $this->task,
            'options' => $fullConfig,
        ]);
    }

    public function delete(): OperationBuilder
    {
        if ($this->queueName === null) {
            throw new \RuntimeException('Queue name is required for delete operation');
        }

        return new OperationBuilder(
            $this->httpClient,
            'DELETE',
            '/api/v1/resources/queues/' . urlencode($this->queueName),
            null
        );
    }

    // ===========================
    // Push Methods
    // ===========================

    public function partition(string $name): static
    {
        $this->partition = $name;
        return $this;
    }

    public function buffer(array $options): static
    {
        $this->bufferOptions = $options;
        return $this;
    }

    public function push(array $payload): PushBuilder
    {
        if ($this->queueName === null) {
            throw new \RuntimeException('Queue name is required for push operation');
        }

        $items = isset($payload[0]) || empty($payload) ? $payload : [$payload];
        $formattedItems = array_map(function (array $item) {
            if (array_key_exists('data', $item)) {
                $payloadValue = $item['data'];
            } elseif (array_key_exists('payload', $item)) {
                $payloadValue = $item['payload'];
            } else {
                $payloadValue = $item;
            }

            $result = [
                'queue' => $this->queueName,
                'partition' => $this->partition,
                'payload' => $payloadValue,
                'transactionId' => $item['transactionId'] ?? Uuid::v7(),
            ];

            if (isset($item['traceId'])) {
                $result['traceId'] = $item['traceId'];
            }

            return $result;
        }, $items);

        return new PushBuilder(
            $this->httpClient,
            $this->bufferManager,
            $this->queueName,
            $this->partition,
            $formattedItems,
            $this->bufferOptions
        );
    }

    // ===========================
    // Consume Configuration
    // ===========================

    public function group(string $name): static
    {
        $this->group = $name;
        return $this;
    }

    public function concurrency(int $count): static
    {
        $this->consumeConcurrency = max(1, $count);
        return $this;
    }

    public function batch(int $size): static
    {
        $this->consumeBatch = max(1, $size);
        return $this;
    }

    public function limit(int $count): static
    {
        $this->consumeLimit = $count;
        return $this;
    }

    public function idleMillis(int $millis): static
    {
        $this->consumeIdleMillis = $millis;
        return $this;
    }

    public function autoAck(bool $enabled): static
    {
        $this->consumeAutoAck = $enabled;
        return $this;
    }

    public function wait(bool $enabled): static
    {
        $this->consumeWait = $enabled;
        return $this;
    }

    public function timeoutMillis(int $millis): static
    {
        $this->consumeTimeoutMillis = $millis;
        return $this;
    }

    public function renewLease(bool $enabled, ?int $intervalMillis = null): static
    {
        $this->consumeRenewLease = $enabled;
        if ($intervalMillis !== null) {
            $this->consumeRenewLeaseIntervalMillis = $intervalMillis;
        }
        return $this;
    }

    public function subscriptionMode(string $mode): static
    {
        $this->consumeSubscriptionMode = $mode;
        return $this;
    }

    public function subscriptionFrom(string $from): static
    {
        $this->consumeSubscriptionFrom = $from;
        return $this;
    }

    public function each(): static
    {
        $this->consumeEach = true;
        return $this;
    }

    // ===========================
    // Consume (callback-based)
    // ===========================

    public function consume(\Closure $handler): ConsumeBuilder
    {
        return new ConsumeBuilder($this->httpClient, $this->queen, $handler, $this->buildConsumeOptions());
    }

    // ===========================
    // High-Level Consumer (rdkafka-style)
    // ===========================

    public function getConsumer(): HighLevelConsumer
    {
        return new HighLevelConsumer($this->httpClient, $this->queen, $this->buildConsumeOptions());
    }

    // ===========================
    // Pop
    // ===========================

    public function pop(): array
    {
        $path = $this->buildPopPath();

        // Pop uses POP_DEFAULTS for autoAck unless explicitly changed
        $effectiveAutoAck = $this->consumeAutoAck !== Defaults::CONSUME_DEFAULTS['autoAck']
            ? $this->consumeAutoAck
            : Defaults::POP_DEFAULTS['autoAck'];

        $params = [
            'batch' => (string) $this->consumeBatch,
            'wait' => $this->consumeWait ? 'true' : 'false',
            'timeout' => (string) $this->consumeTimeoutMillis,
        ];

        if ($this->group !== null) {
            $params['consumerGroup'] = $this->group;
        }
        if ($this->namespace !== null) {
            $params['namespace'] = $this->namespace;
        }
        if ($this->task !== null) {
            $params['task'] = $this->task;
        }
        if ($effectiveAutoAck) {
            $params['autoAck'] = 'true';
        }
        if ($this->consumeSubscriptionMode !== null) {
            $params['subscriptionMode'] = $this->consumeSubscriptionMode;
        }
        if ($this->consumeSubscriptionFrom !== null) {
            $params['subscriptionFrom'] = $this->consumeSubscriptionFrom;
        }

        $query = http_build_query($params);
        $affinityKey = $this->getAffinityKey();
        $result = $this->httpClient->get("{$path}?{$query}", $this->consumeTimeoutMillis + 5000, $affinityKey);

        if (!$result || !isset($result['messages'])) {
            return [];
        }

        return array_filter($result['messages'], fn($msg) => $msg !== null);
    }

    // ===========================
    // Buffer
    // ===========================

    public function flushBuffer(): void
    {
        if ($this->queueName === null) {
            throw new \RuntimeException('Queue name is required for buffer flush');
        }
        $queueAddress = "{$this->queueName}/{$this->partition}";
        $this->bufferManager->flushBuffer($queueAddress);
    }

    // ===========================
    // DLQ
    // ===========================

    public function dlq(?string $consumerGroup = null): DLQBuilder
    {
        if ($this->queueName === null) {
            throw new \RuntimeException('Queue name is required for DLQ operations');
        }
        return new DLQBuilder($this->httpClient, $this->queueName, $consumerGroup, $this->partition);
    }

    // ===========================
    // Private helpers
    // ===========================

    private function buildConsumeOptions(): array
    {
        return [
            'queue' => $this->queueName,
            'partition' => $this->partition !== 'Default' ? $this->partition : null,
            'namespace' => $this->namespace,
            'task' => $this->task,
            'group' => $this->group,
            'concurrency' => $this->consumeConcurrency,
            'batch' => $this->consumeBatch,
            'limit' => $this->consumeLimit,
            'idleMillis' => $this->consumeIdleMillis,
            'autoAck' => $this->consumeAutoAck,
            'wait' => $this->consumeWait,
            'timeoutMillis' => $this->consumeTimeoutMillis,
            'renewLease' => $this->consumeRenewLease,
            'renewLeaseIntervalMillis' => $this->consumeRenewLeaseIntervalMillis,
            'subscriptionMode' => $this->consumeSubscriptionMode,
            'subscriptionFrom' => $this->consumeSubscriptionFrom,
            'each' => $this->consumeEach,
        ];
    }

    private function buildPopPath(): string
    {
        if ($this->queueName !== null) {
            if ($this->partition !== 'Default') {
                return "/api/v1/pop/queue/{$this->queueName}/partition/{$this->partition}";
            }
            return "/api/v1/pop/queue/{$this->queueName}";
        }

        if ($this->namespace !== null || $this->task !== null) {
            return '/api/v1/pop';
        }

        throw new \RuntimeException('Must specify queue, namespace, or task for pop operation');
    }
}
